packages/server/postgres/utils/checkEqBase.ts

```diff
@@ -0,0 +1,164 @@
+import {RTable, TableSchema} from '../../database/stricterR'
+import areEqual from 'fbjs/lib/areEqual'
+import getPg from '../getPg'
+
+interface DBDoc {
+  id: string
+  [key: string]: any
+}
+interface RethinkDoc extends DBDoc {}
+interface PGDoc extends DBDoc {}
+
+interface IError {
+  [key: string]:
+    | {
+        error: string[]
+        rethinkRow: Partial<RethinkDoc>
+        pgRow: Partial<PGDoc>
+      }
+    | number
+    | undefined
+    | string[]
+  rethinkRecordsCompared: number
+  foundErrors: number
+  missingPgRows: string[]
+  extraPgRows: string[]
+  rethinkRowCount?: number
+  pgRowCount?: number
+}
+
+function getPairNeFields(
+  rethinkRow: RethinkDoc,
+  pgRow: PGDoc,
+  alwaysDefinedFields: string[],
+  maybeUndefinedFieldsDefaultValues: {[key: string]: any}
+): string[] {
+  const neFields = [] as string[]
+
+  for (const f of alwaysDefinedFields) {
+    const [rethinkValue, pgValue] = [rethinkRow[f], pgRow[f]]
+    if (!areEqual(rethinkValue, pgValue)) {
+      neFields.push(f)
+    }
+  }
+  for (const [f, defaultValue] of Object.entries(maybeUndefinedFieldsDefaultValues)) {
+    const [rethinkValue, pgValue] = [rethinkRow[f], pgRow[f]]
+    if (rethinkValue !== undefined) {
+      if (!areEqual(rethinkValue, pgValue)) {
+        neFields.push(f)
+      }
+    } else {
+      if (!areEqual(pgValue, defaultValue)) {
+        neFields.push(f)
+      }
+    }
+  }
+
+  return neFields
+}
+
+function addNeFieldsToErrors(
+  errors: IError,
+  neFields: string[],
+  rethinkRow: RethinkDoc,
+  pgRow: PGDoc,
+  rowId: string
+): void {
+  const rethinkNeFields = {}
+  const pgNeFields = {}
+  for (const neField of neFields) {
+    rethinkNeFields[neField] = rethinkRow[neField]
+    pgNeFields[neField] = pgRow[neField]
+  }
+  errors[rowId] = {
+    error: neFields,
+    rethinkRow: rethinkNeFields,
+    pgRow: pgNeFields
+  }
+}
+
+const getExtraPgRowIds = async (tableName: string): Promise<string[]> => {
```

Comment:
not sure i understand this part!

making sure rethinkdb & PG have the same number of rows is probably the final test (after making sure there are no missing rows in pg & all the checked rows are equal). if any of those other errors are present, we don't need to check equality.


if that's the case, could we just write `SELECT COUNT(*)` for PG & `r.table('Team').count()` for rethinkdb & compare the number? if we do that could we remove the `eqChecked` column?

Reply:
we could just do an aggregate count on both tables. what i was trying to avoid was a scenario in which we happened to have more pg rows than rethink rows, and then it's up to the caller to figure out what these extra rows are from there.

having the extra field is more efficient in terms of time and i think it's a more complete solution to return any possible extra pg rows straight away, but i also agree it may be premature optimization.

Reply:
ahhh i see!
performance is definitely important, but since we'll only run these queries a handful of times, I'd opt for a pattern that we can repeat as quickly as possible for the remaining tables over query performance. 

If I understand the tradeoff, if we _just_ do a count, we could avoid creating & destroying an index & the query would only have to read instead of upsert. On the negative, if we wanted to see which rows only existed in PG, we'd have fetch all the ids in rethinkdb and perform a `NOT IN (...)` scan that would approach O(n^2).

for the sake of speed, I'd say we:
- remove the eqChecked index
- convert the upsert to a read query
- perform a count on both DBs & if they are different, add an error line that shows the difference

if we do come across that dreadful case where we have more docs in pg than rethinkdb we can always write a new PR!

Reply:
> I'd opt for a pattern that we can repeat as quickly as possible for the remaining tables over query performance.

yeah that's a heuristic i wasn't thinking of in terms of tradeoffs, i see what you mean

Reply:
done!